tree/treeplayer/test/array.cxx

```diff
@@ -66,3 +67,138 @@ TEST(TTreeReaderArray, MultiReaders) {
       EXPECT_DOUBLE_EQ(Double[i], trDouble5[i]);
    }
 }
+
+TEST(TTreeReaderArray, Iterator) {
+   // test TTreeReaderArray's nested iterator type
+
+   // create input TTree
+   TTree tree("TTreeReaderArrayTree", "In-memory test tree");
+   std::array<int, 6> values{0, 1, 2, 3, 4, 5};
+   int n = 6;
+   tree.Branch("n", &n);
+   tree.Branch("arr", &values, "arr[n]/I");
+   tree.Fill();
+   n = 0;
+   tree.Fill();
+   tree.ResetBranchAddresses();
+
+   // create TTreeReader
+   TTreeReader r(&tree);
+   TTreeReaderArray<int> arr(r, "arr");
+
+   // tests on full entry
+   r.Next();
+
+   // begin and end, comparisons
+   EXPECT_EQ(arr.begin(), arr.begin());
+   EXPECT_EQ(arr.end(), arr.end());
+   EXPECT_NE(arr.begin(), arr.end());
+
+   // copy/move ctors
+   auto it = arr.begin();
+   const auto it2(it);
+   auto it3(std::move(it));
+
+   // increments, decrements, copy/move assignments
+   EXPECT_EQ(++it, arr.begin() + 1);
+   it = it3++;
+   EXPECT_EQ(it += 1, it3);
+
+   EXPECT_EQ((arr.begin() + 2) - 1, --(arr.begin() + 2));
+   it = std::move(it3);
+   auto it4 = it--;
+   EXPECT_EQ(it4 -= 1, it);
+
+   // TODO add operator+(int, iterator)
+   // 3 + it;
+
+   // comparisons
+   auto e = arr.end();
+   auto b = arr.begin();
+   EXPECT_GT(e, b);
+   EXPECT_LT(b, b + 2);
+   EXPECT_LE(b + 1, b + 1);
+   EXPECT_GE(b + 1, b + 1);
+
+   // difference
+   const auto length = arr.end() - arr.begin();
```

Comment:
Can you use the actual type here, as that is part of the iterator pattern? Just to make sure that if people rely on it being `std::ptrdiff_t` it actually compiles without a warning.

Reply:
Done